kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/blockchain/BlockchainSourceImpl.java

```diff
@@ -156,48 +187,78 @@ public void onError(final Exception e) {
                         }
                     }
                 });
-                Log.d(TAG, "getCurrentBalance onError: " + e.getMessage());
+                Log.d(TAG, "getBalance onError: " + e.getMessage());
             }
         });
     }
 
-    private void setBalance(int balanceValue) {
-        balance.setValue(balanceValue);
-        local.setBalance(balanceValue);
+    @VisibleForTesting
+    void setBalance(final Balance balanceObj) {
+        BalanceUpdate balanceUpdate = balance.getValue();
+        // if the values are not equals so we need to update,
+        // no need to update for equal values.
+        if (balanceUpdate.getAmount().compareTo(balanceObj.value()) != 0) {
+            Log.d(TAG, "setBalance: Balance changed, should get update");
+            balanceUpdate.setAmount(balanceObj.value());
+            balance.postValue(balanceUpdate);
+            local.setBalance(balanceObj.value().intValue());
+        }
     }
 
     @Override
-    public void addBalanceObserver(@NonNull Observer<Integer> observer) {
+    public void addBalanceObserver(@NonNull Observer<BalanceUpdate> observer) {
         balance.addObserver(observer);
         observer.onChanged(balance.getValue());
     }
 
     @Override
-    public void removeBalanceObserver(@NonNull Observer<Integer> observer) {
+    public void addBalanceObserverAndStartListen(@NonNull Observer<BalanceUpdate> observer) {
+        addBalanceObserver(observer);
+        Log.d(TAG, "addBalanceObserverAndStartListen: " + balanceObserversCount.get());
+        if (balanceObserversCount.getAndIncrement() == 0) {
+            startBalanceListener();
+        }
+    }
+
+    private void startBalanceListener() {
+        Log.d(TAG, "startBalanceListener: ");
+        balanceRegistration = account.blockchainEvents()
+            .addBalanceListener(new EventListener<Balance>() {
+                @Override
+                public void onEvent(Balance data) {
+                    setBalance(data);
+                }
+            });
+    }
+
+    @Override
+    public void removeBalanceObserver(@NonNull Observer<BalanceUpdate> observer) {
+        Log.d(TAG, "removeBalanceObserver: ");
         balance.removeObserver(observer);
     }
 
     @Override
-    public void createTrustLine() {
-        account.activate().run(new ResultCallback<Void>() {
-            @Override
-            public void onResult(Void result) {
-                Log.d(TAG, "createTrustLine onResult");
-            }
+    public void removeBalanceObserverAndStopListen(@NonNull Observer<BalanceUpdate> observer) {
+        removeBalanceObserver(observer);
+        decrementBalanceCount();
+    }
 
-            @Override
-            public void onError(Exception e) {
-                Log.d(TAG, "createTrustLine onError");
-            }
-        });
+    private void decrementBalanceCount() {
+        if (balanceObserversCount.get() > 0 &&
+            balanceObserversCount.decrementAndGet() == 0) {
```

Comment:
checking both conditions is not an atomic operation, you should use `compareAndSet(1, 0)`, but anyway here you should lock the entire section, then updating the count and start/stop the registration at once